lib/spack/spack/installer.py

```diff
@@ -613,6 +613,42 @@ def package_id(pkg):
     return "{0}-{1}-{2}".format(pkg.name, pkg.version, pkg.spec.dag_hash())
 
 
+def update_kwargs_from_args(args, kwargs):
```

Comment:
I feel like this is the wrong place for this function if the aim was to have the spack core module not dependent on the cli. Here we assume knowledge on which arguments are passed in an `argparse` namespace.

Reply:
Yeah, I tried to explain this in the commit message 89177317cc2b50e2845bf968507c8afa3a0ab0a5:

> The way I see it update_kwargs_from_args could have been a constructor
> for a PackageInstallArgs object, and they could have a named constructor
> like PackageInstallArgs::from_cli_args(args). So it makes sense to move
> this function to spack.installer, and with that the last dependency of
> spack.environment on spack.cmd is gone, as it should be.

Reply:
I could move it out of installer.py, but I don't know where it should go instead :thinking:

Reply:
Ok, I see that commit. My opinion is that the right solution is to have:

https://github.com/spack/spack/blob/b22728d55c126a0034f70c7331b6fabc4d6ad095/lib/spack/spack/environment.py#L1575

and similar functions not having the `args` argument. It should be on the caller of the interface to prepare the `install_args` appropriately.

Reply:
Looking at:

https://github.com/spack/spack/blob/b22728d55c126a0034f70c7331b6fabc4d6ad095/lib/spack/spack/cmd/install.py#L197

it seems this is doable. Besides, `update_kwargs_from_args` is called _internally_ from `Environment.install_specs` which receives directly the `cli_args` and externally in the function above. So why not setting everything in `cmd/install.py` and just pass a dictionary after a single call to the function?

Reply:
I've removed `args` from `spack.environment.install_spec(...)` and `spack.environment.install_all(...)` and combined the `update_kwargs_from_args` in `spack.cmd.install`. That seems like a better solution indeed.

Reply:
That's exactly what I had in mind. Thanks!